mesh/mesh.cpp

```diff
@@ -8247,6 +8247,188 @@ void Mesh::PrintVTK(std::ostream &out)
+         }
+         out << '\n';
+      }
+   }
+   out<<"</DataArray>"<<std::endl;
+   out<<"</Points>"<<std::endl;
+
+   out<<"<Cells>"<<std::endl;
+   out<<"<DataArray type=\"Int32\" Name=\"connectivity\" format=\"ascii\">"<<std::endl;
+   //connectivity
+   std::vector<int> offset;
+   int coff=0;
+   np = 0;
+   for (int i = 0; i < GetNE(); i++)
+   {
+      Geometry::Type geom = GetElementBaseGeometry(i);
+      int nv = Geometries.GetVertices(geom)->GetNPoints();
+      RefG = GlobGeometryRefiner.Refine(geom, ref, 1);
+      Array<int> &RG = RefG->RefGeoms;
+
+      for (int j = 0; j < RG.Size(); )
+      {
+         //out << nv;
+         coff=coff+nv;
+         offset.push_back(coff);
+
+         for (int k = 0; k < nv; k++, j++)
+         {
+            out << ' ' << np + RG[j];
+         }
+         out << '\n';
+      }
+      np += RefG->RefPts.GetNPoints();
+   }
+   out<<"</DataArray>"<<std::endl;
+
+   out<<"<DataArray type=\"Int32\" Name=\"offsets\" format=\"ascii\">"<<std::endl;
+   //offsets
+   for (size_t ii=0; ii<offset.size(); ii++) { out<<offset[ii]<<std::endl;}
+   out<<"</DataArray>"<<std::endl;
+   out<<"<DataArray type=\"UInt8\" Name=\"types\" format=\"ascii\">"<<std::endl;
+   //cell types
+   for (int i = 0; i < GetNE(); i++)
+   {
+      Geometry::Type geom = GetElementBaseGeometry(i);
+      int nv = Geometries.GetVertices(geom)->GetNPoints();
+      RefG = GlobGeometryRefiner.Refine(geom, ref, 1);
+      Array<int> &RG = RefG->RefGeoms;
+      int vtk_cell_type = 5;
+
+      switch (geom)
+      {
+         case Geometry::POINT:        vtk_cell_type = 1;   break;
+         case Geometry::SEGMENT:      vtk_cell_type = 3;   break;
+         case Geometry::TRIANGLE:     vtk_cell_type = 5;   break;
+         case Geometry::SQUARE:       vtk_cell_type = 9;   break;
+         case Geometry::TETRAHEDRON:  vtk_cell_type = 10;  break;
+         case Geometry::CUBE:         vtk_cell_type = 12;  break;
+         case Geometry::PRISM:        vtk_cell_type = 13;  break;
+         default:
+            MFEM_ABORT("Unrecognized VTK element type \"" << geom << "\"");
+            break;
+      }
+
+      for (int j = 0; j < RG.Size(); j += nv)
+      {
+         out << vtk_cell_type << '\n';
+      }
+   }
+   out<<"</DataArray>"<<std::endl;
+   out<<"</Cells>"<<std::endl;
+
+   out<<"<CellData Scalars=\"material\">"<<std::endl;
+   out<<"<DataArray type=\"Int32\" Name=\"material\" format=\"ascii\">"<<std::endl;
+   for (int i = 0; i < GetNE(); i++)
+   {
+      Geometry::Type geom = GetElementBaseGeometry(i);
+      int nv = Geometries.GetVertices(geom)->GetNPoints();
+      RefG = GlobGeometryRefiner.Refine(geom, ref, 1);
+      int attr = GetAttribute(i);
+      for (int j = 0; j < RefG->RefGeoms.Size(); j += nv)
+      {
+         out << attr << '\n';
+      }
+   }
+   out<<"</DataArray>"<<std::endl;
+
+   if (Dim > 1)
+   {
+      out<<"<DataArray type=\"Int32\" Name=\"element_coloring\" format=\"ascii\">"<<std::endl;
```

Comment:
Some formatting, like this example, is still off.

Reply:
All streams should be fine now.